src/coreComponents/mesh/utilities/MeshMapUtilities.hpp

```diff
@@ -126,103 +131,167 @@ inline localIndex size1( ArrayOfSetsView< T const > const & map, localIndex cons
+
+  // Calculate the number of entries in each sub-array
+  array1d< localIndex > elemCounts( numObjects );
+  elemCounts.setValues< serialPolicy >( overAlloc );
+  elementRegionManager.forElementSubRegionsComplete< CellElementSubRegion >( [&]( localIndex const,
+                                                                                  localIndex const,
+                                                                                  ElementRegionBase const &,
+                                                                                  CellElementSubRegion const & subRegion )
   {
-    size += map.sizeOfArray( i );
-  }
-  return size;
-}
+    auto const elemToObject = cellToObjectGetter( subRegion );
+    forAll< parallelHostPolicy >( subRegion.size(), [elemCounts = elemCounts.toView(),
+                                                     elemToObject]( localIndex const ei )
+    {
+      for( localIndex a = 0; a < size1( elemToObject, ei ); ++a )
+      {
+        localIndex const objID = elemToObject( ei, a );
+        RAJA::atomicInc< parallelHostAtomic >( &elemCounts[ objID ] );
+      }
+    } );
+  } );
 
-/**
- * @copydoc numElements(arrayView1d< arrayView1d< T const > const > const &)
- */
-template< typename T >
-GEOSX_HOST_DEVICE
-inline localIndex numElements( ArrayOfSetsView< T const > const & map )
-{
-  localIndex size = 0;
-  for( localIndex i = 0; i < map.size(); ++i )
+  // Allocate memory
+  using ElemID = std::tuple< localIndex, localIndex, localIndex >;
+  ArrayOfArrays< ElemID > elemList;
+  elemList.resizeFromCapacities< parallelHostPolicy >( numObjects, elemCounts.data() );
+
+  // Populate map of tuples in parallel
+  elementRegionManager.forElementSubRegionsComplete< CellElementSubRegion >( [&]( localIndex const er,
+                                                                                  localIndex const esr,
+                                                                                  ElementRegionBase const &,
+                                                                                  CellElementSubRegion const & subRegion )
   {
-    size += map.sizeOfSet( i );
-  }
-  return size;
+    auto const elemToObject = cellToObjectGetter( subRegion );
+    forAll< parallelHostPolicy >( subRegion.size(), [elemList = elemList.toView(),
+                                                     elemToObject, er, esr]( localIndex const ei )
+    {
+      for( localIndex a = 0; a < size1( elemToObject, ei ); ++a )
+      {
+        localIndex const objIndex = elemToObject( ei, a );
+        elemList.emplaceBackAtomic< parallelHostAtomic >( objIndex, er, esr, ei );
+      }
+    } );
+  } );
+
+  // Finally, split arrays-of-tuples into separate arrays
+  // TODO: this resizing ends up scanning elemCounts 3 times, need new ArrayOfArrays API?
+  toElementRegionList.resizeFromCapacities< parallelHostPolicy >( numObjects, elemCounts.data() );
```

Comment:
Couldn't you do

``` c++
toElementRegionList.resizeFromCapacities< parallelHostPolicy >( numObjects, elemCounts.data() );
toElementSubRegionList = toElementRegionList;
toElementList = toElementRegionList;
```
I'm pretty sure that would be more efficient, but I doubt it really matters.

Edit:
Actually it will most likely be slower because `operator=` basically involves a `memcpy` for each of in inner arrays. If this pattern becomes prevalent we could easily add a function `duplicateSparsity` that would do what we want. As my naming suggests I think this would be particularly useful for the matrix classes.